Assets/MixedRealityToolkit/_Core/Devices/OpenVR/OpenVRDeviceManager.cs

```
@@ -1,11 +1,158 @@
+
+        /// <summary>
+        /// Tracking states returned from the InputTracking state tracking manager
+        /// </summary>
+        private List<XRNodeState> nodeStates = new List<XRNodeState>();
+
+        /// <inheritdoc/>
+        public override IMixedRealityController[] GetActiveControllers()
+        {
+            return activeControllers.Values.ToArray();
+        }
+
+        public override void Enable()
+        {
+            InputTracking.nodeAdded += InputTracking_nodeAdded;
+            InputTracking.nodeRemoved += InputTracking_nodeRemoved;
+            InputTracking.trackingAcquired += InputTracking_trackingAcquired;
+            InputTracking.trackingLost += InputTracking_trackingLost;
+        }
+
+        public override void Update()
+        {
+            GenericOpenVRController controller;
+            //This is so gosh-darn awful
+            InputTracking.GetNodeStates(nodeStates);
+            for (int i = 0; i < nodeStates.Count; i++)
+            {
+                if (IsNodeTypeSupported(nodeStates[i]) && activeControllers.ContainsKey(nodeStates[i].nodeType))
+                {
+                    controller = activeControllers[nodeStates[i].nodeType] as GenericOpenVRController;
+                    controller.UpdateController(nodeStates[i]);
+                }
+            }
+        }
+
+        #region Unity InteractionManager Events
+
+        private void InputTracking_nodeAdded(XRNodeState obj)
+        {
+            if (IsNodeTypeSupported(obj))
+            {
+                GetOrAddController(obj);
+                Debug.Log($"Added node for [{obj.nodeType}]");
+            }
+        }
+
+        private void InputTracking_trackingAcquired(XRNodeState obj)
+        {
+            if (IsNodeTypeSupported(obj))
+            {
+                var controller = GetOrAddController(obj);
+                InputSystem?.RaiseSourceDetected(controller?.InputSource, controller);
```

Why would a controller be null here?

---

Belt and braces. Copied from the WMR implementation.

---

Let's raise an error or assert that it's not null.

There's not really any reason why it would be null here.

---

Fair enough.   But will do a separate PR to address both at the same time so we don't get confused :D